modules/dbparser/tests/test_parsers.c

```diff
@@ -74,56 +56,72 @@ _invoke_parser(gboolean (*parser)(guint8 *str, gint *len, const gchar *param, gp
   return result;
 }
 
-static void
-assert_parser_success(gboolean (*parser)(guint8 *str, gint *len, const gchar *param, gpointer state,
-                                         RParserMatch *match), const gchar *str, gpointer param, gpointer state, const gchar *expected_string)
+typedef struct _parser_test_param
 {
-  gchar *result_string = NULL;
-  gboolean result;
+  gchar *str;
+  gpointer param;
+  gchar *expected_string;
+  gboolean expected_result;
+} ParserTestParam;
 
-  result = _invoke_parser(parser, str, param, state, &result_string);
-  assert_true(result, "Mismatching parser result");
-  assert_string(result_string, expected_string, "Mismatching parser result");
-  g_free(result_string);
+ParameterizedTestParameters(parser, test_string_parser)
+{
+  static ParserTestParam parser_params[] =
+  {
+    {.str = "foo", .param = NULL, .expected_string = "foo", .expected_result = TRUE},
+    {.str = "foo bar", .param = NULL, .expected_string = "foo", .expected_result = TRUE},
+    {.str = "foo123 bar", .param = NULL, .expected_string = "foo123", .expected_result = TRUE},
+    {.str = "foo{}", .param = NULL, .expected_string = "foo", .expected_result = TRUE},
+    {.str = "foo[]", .param = NULL, .expected_string = "foo", .expected_result = TRUE},
+    {.str = "foo", .param = "X", .expected_string = "foo", .expected_result = TRUE},
+    {.str = "foo=bar", .param = "=", .expected_string = "foo=bar", .expected_result = TRUE},
+    {.str = "", .param = NULL, .expected_string = NULL, .expected_result = FALSE},
+  };
+
+  return cr_make_param_array(ParserTestParam, parser_params,
+                             sizeof(parser_params) / sizeof(parser_params[0]));
 }
 
-static void
-assert_parser_failure(gboolean (*parser)(guint8 *str, gint *len, const gchar *param, gpointer state,
-                                         RParserMatch *match), const gchar *str, gpointer param, gpointer state)
+ParameterizedTest(ParserTestParam *param, parser, test_string_parser)
 {
+  gchar *result_string = NULL;
   gboolean result;
 
-  result = _invoke_parser(parser, str, param, state, NULL);
-  assert_false(result, "Mismatching parser result");
+  result = _invoke_parser(r_parser_string, param->str, param->param, NULL, &result_string);
+  if (param->expected_result == TRUE)
+    {
+      cr_assert(result, "Mismatching parser result");
+      cr_assert_str_eq(result_string, param->expected_string, "Mismatching parser result");
```

Comment:
I'm afraid this hides the strings in case of error. We should either not provide an error string, or print the underlying strings similarly to the bottom of the file as %s %s.

Reply:
I added the result and expected strings to printout.